edge-hub/core/src/Microsoft.Azure.Devices.Edge.Hub.Http/controllers/RegistryController.cs

```diff
@@ -0,0 +1,735 @@
+                }
+                catch (Exception ex)
+                {
+                    Events.BadRequest(ex.Message);
+                    await this.SendResponseAsync(HttpStatusCode.BadRequest, this.FormatErrorResponseMessage(ex));
+                    return;
+                }
+
+                actorDeviceId = WebUtility.UrlDecode(actorDeviceId);
+
+                if (!ControllerHelper.TryGetTargetDeviceId(requestData.AuthChain, out string targetDeviceId))
+                {
+                    Events.InvalidRequestAuthChain(requestData.AuthChain);
+                    await this.SendResponseAsync(HttpStatusCode.BadRequest, this.FormatErrorResponseMessage($"Invalid request auth chain {requestData.AuthChain}."));
+                    return;
+                }
+
+                IHttpRequestAuthenticator authenticator = await this.authenticatorGetter;
+                HttpAuthResult authResult = await authenticator.AuthenticateAsync(actorDeviceId, Option.Some(Constants.EdgeHubModuleId), this.HttpContext);
+
+                if (!authResult.Authenticated)
+                {
+                    Events.AuthenticateFail(actorDeviceId, Constants.EdgeHubModuleId);
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
+                }
+
+                Events.Authenticated(actorDeviceId, Constants.EdgeHubModuleId);
+                // Check that the actor device is authorized to act OnBehalfOf the target
+                IEdgeHub edgeHub = await this.edgeHubGetter;
+                IDeviceScopeIdentitiesCache identitiesCache = edgeHub.GetDeviceScopeIdentitiesCache();
+
+                (bool authorized, string authChain) = await this.AuthorizeActorAsync(identitiesCache, actorDeviceId, Constants.EdgeHubModuleId, targetDeviceId);
+                if (!authorized)
+                {
+                    await this.SendResponseAsync(HttpStatusCode.Unauthorized);
+                    return;
+                }
+
+                Events.Authorized(actorDeviceId, Constants.EdgeHubModuleId, targetDeviceId, authChain);
+                string edgeDeviceId = edgeHub.GetEdgeDeviceId();
+                requestData.AuthChain = $"{authChain}";
+                HttpResponseMessage responseMessage = await this.apiClient.DeleteModuleOnBehalfOfAsync(edgeDeviceId, requestData);
+                await this.SendResponseAsync(responseMessage);
+                Events.CompleteRequest(edgeDeviceId, targetDeviceId, requestData.ModuleId, authChain);
+            }
+            catch (Exception ex)
+            {
+                await this.SendResponseAsync(HttpStatusCode.InternalServerError, this.FormatErrorResponseMessage(ex));
+            }
+        }
+
+        async Task<(bool, string)> AuthorizeActorAsync(IDeviceScopeIdentitiesCache identitiesCache, string actorDeviceId, string actorModuleId, string targetId)
```

Comment:
>AuthorizeActorAsync [](start = 35, length = 19)

seems identitical to the one in DeviceScopeController, maybe put it in HttpRequestAuthenticator to share across both classes.